src/emmy/expression/analyze.cljc

```diff
@@ -81,25 +82,53 @@
       :else (compare v w))))
 
 (defn monotonic-symbol-generator
-  "Returns a function which generates a sequence of symbols with the given
-  `prefix` with the property that later symbols will sort after earlier symbols.
-
-  This is important for the stability of the simplifier. (If we just used
-  `clojure.core/gensym`, then a temporary symbol like `G__1000` will sort
-  earlier than `G__999`. This will trigger errors at unpredictable times,
-  whenever `clojure.core/gensym` returns two symbols that cross an
-  order-of-magnitude boundary.)"
-  [prefix]
-  (let [count (atom -1)]
-    (fn [] (symbol
-           #?(:clj
-              (format "%s%016x" prefix (swap! count inc))
-
-              :cljs
-              (let [i (swap! count inc)
-                    suffix (-> (.toString i 16)
-                               (.padStart 16 "0"))]
-                (str prefix suffix)))))))
+  "Called with no arguments, produces a function mapping a string prefix
+   to a generated symbol with a four-digit suffix which increments
+   with each call, providing a stream of unique symbols. If the returned
+   function is called without arguments, a default prefix of \"_\" is
+   used (but see below).
+
+   May be called with one integer argument to set the size of the
+   suffix field in digits.
+
+   Supplying yet one more string argument changes the default prefix.
+   ```
+   (def g (monotonic-symbol-generator))
+   (take 5 (repeatedly #(g \"a\")))
+   (take 5 (repeatedly g))
+   ;; (a0000 a0001 a0002 a0003 a0004)
+   ;; (_0005 _0006 _0007 _0008 _0009)
+
+
+   (def h (monotonic-symbol-generator 2))
+   (take 5 (repeatedly #(h \"b\")))
+   (take 5 (repeatedly h))
+   ;; (b00 b01 b02 b03 b04)
+   ;; (_05 _06 _07 _08 _09)
+
+   (def j (monotonic-symbol-generator 3 \"x\"))
+   (take 5 (repeatedly #(j \"a\")))
+   (take 5 (repeatedly j))
+   ;; (a000 a001 a002 a003 a004)
+   ;; (x005 x006 x007 x008 x009)
+   ```
+   "
+  ([]
+   (monotonic-symbol-generator 4 "_"))
+  ([size]
+   (monotonic-symbol-generator size "_"))
+  ([size prefix]
```

Comment:
Should we assert that the prefix doesn't start with a number (making sure that we get valid symbols out), or is this too pedantic?